File: go/LeetCode/HeapSort.go
package main

func main() {
	nums := []int{4,1,6,5,9}
	heapSort(nums)
	for _,v :=range nums{
		print(v)
	}
}

func heapSort(nums []int) []int {
	len := len(nums)
	for i:=len/2;i>=0;i-- {
		sink(nums, i, len)
	}
	for len>1 {
		len -=1
		swap(nums, 0, len)
		sink(nums, 0 ,len)
	}
	return nums
}

func sink(nums []int, k int, len int)  {
	for 2*k+1<len {
		j:=2*k+1
		for j+1<len && nums[j+1]>nums[j] {
			j++
		}
		if nums[k]> nums[j] {
			break
		}
		swap(nums, j, k)
		k=j
	}
}

func swap(nums []int, i int, j int)  {
	temp := nums[i]
	nums[i] = nums[j]
	nums[j] = temp
}

File: python3/LeetCode/HeapSort.py
from typing import List


class Solution:
    def heapSort(self, nums:List[int]):
        length = len(nums)
        for i in range(length//2, -1, -1):
            self.sink(nums, i, length)

        while length>1:
            length-=1
            self.swap(nums, 0, length)
            self.sink(nums, 0, length)



    def sink(self, nums:List[int], k:int, len:int):
        while 2*k+1<len:
            j = 2*k+1
            while j+1<len and nums[j+1]>nums[j]:
                j+=1
            if nums[k]> nums[j]:
                break
            self.swap(nums, j, k)
            k=j

    def swap(self, nums: List[int], i: int, j: int):
                temp = nums[i]
                nums[i] = nums[j]
                nums[j] = temp

if __name__ == '__main__':
    nums = [5,3,6,1,8]
    Solution().heapSort(nums)
    print(nums)